### PR TITLE
Update Dockerfile

### PR DESCRIPTION
**Pull request by klaemo**
Hey there,
this PR brings the Dockerfile up to speed.

It allows developers to "quickly" (technically, it still takes a while :D) build and run couchdb master with the following commands:

```
docker build -t couchdb-master
docker run -i -t --rm couchdb-master
```

It'll build the documentation and fauxton as well, that's the reason for all the dependencies.
I intent to make further improvements in the future. For example, the haproxy integration doesn't work yet (no idea why, it works in my other repo) and I think we could even provide automated builds through Docker Hub, so people can just download and run a container based on the master branch for quick testing and the like (basically what I'm doing [here](https://github.com/klaemo/docker-couchdb/tree/master/2.0-dev)).

Feedback welcome :)

### CODE REVIEW

**Comment by kxepal** [comment]
Is it possible to pull Dockerfile from `couchdb-docker`? Like `./configure --with docker && docker ...` and here we go?

**Comment by klaemo** [comment]
It's possible, but I think that involves cloning the couchdb repo once more inside of the container.
Essentially, there are two ways I can think of:
1. Have the Dockerfile along side the src, `COPY` the (already cloned) src into the container and build that beast. It pretty much acts like a Makefile. In that case having the Dockerfile in a different repo wouldn't make much sense. The upside is cloning the repo only once and being aware of local changes the developer might have made.
2. Pull in the Dockerfile from `couchdb-docker`, clone the repo inside of the container and build. Downside: having to clone twice and it doesn't take local src changes into account because it always pulls from the remote (that's how https://github.com/klaemo/docker-couchdb/tree/master/2.0-dev works atm)

A combination of 1 and 2 is possible, but it seems weird to have the Dockerfile in a different repo if it depends on the source code being there.

**Comment by robertkowalski** [comment]
@klaemo just wanted to test before +1 it, i get:

```
(19:20:25) [robert@tequila-work] ~/apache/couchdb $ docker run -i -t --rm couchdb-master
[ * ] Setup environment ... ok
[ * ] Ensure CouchDB is built ... ok
[ * ] Prepare configuration files ... ok
[ * ] Start node node1 ... ok
[ * ] Start node node2 ... ok
[ * ] Start node node3 ... ok
[ * ] Check node at http://127.0.0.1:15984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:25984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:35984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:15984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:25984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:35984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:15984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:25984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:35984/ ... ^Cok

```

**Comment by klaemo** [comment]
Weird, here's what I get:

```
$ docker run -i -t --rm couchdb-master
[ * ] Setup environment ... ok
[ * ] Ensure CouchDB is built ... ok
[ * ] Prepare configuration files ... ok
[ * ] Start node node1 ... ok
[ * ] Start node node2 ... ok
[ * ] Start node node3 ... ok
[ * ] Check node at http://127.0.0.1:15984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:25984/ ... ok
[ * ] Check node at http://127.0.0.1:35984/ ... ok
[ * ] Check node at http://127.0.0.1:15984/ ... ok
[ * ] Running cluster setup ... ok
[ * ] Developers cluster is set up at http://127.0.0.1:15984.
Admin username: root
Password: kw9ZYQcx
Time to hack! ... 
```

I'll take another look at it, though.

**Comment by snowch** [comment]
Seems to be working ok for me:

```
vagrant@vagrant-ubuntu-trusty-64:~/Repos/couchdb-klaemo$ docker run -i -t -P --rm couchdb-master --with-haproxy
[ * ] Setup environment ... ok
[ * ] Ensure CouchDB is built ... ok
[ * ] Prepare configuration files ... ok
[ * ] Start node node1 ... ok
[ * ] Start node node2 ... ok
[ * ] Start node node3 ... ok
[ * ] Check node at http://127.0.0.1:15984/ ... failed: [Errno socket error] [Errno 111] Connection refused
[ * ] Check node at http://127.0.0.1:25984/ ... ok
[ * ] Check node at http://127.0.0.1:35984/ ... ok
[ * ] Check node at http://127.0.0.1:15984/ ... ok
[ * ] Running cluster setup ... ok
[ * ] Developers cluster is set up at http://127.0.0.1:15984.
Admin username: root
Password: Dog2iOiA
Time to hack! ... 
```
I also managed to run OK with haproxy.  To start the container then access haproxy in the container from the docker host:

```
host$ COUCH_INSTANCE_ID=$(docker run -d -P couchdb-master --with-haproxy)
host$ COUCH_INSTANCE_IP=$(docker inspect --format='{{.NetworkSettings.IPAddress}}' $COUCH_DOCKER_ID)
host$ firefox ${COUCH_INSTANCE_IP}:5984
```

**Comment by snowch** [comment]
Although haproxy was running, it wasn't logging.  To get haproxy logging to /var/log/haproxy, I added the following to the global section of /usr/src/couchdb/rel/haproxy.cfg:

```
       log /dev/log    local0
       log /dev/log    local1 notice
       chroot /var/lib/haproxy
       user haproxy
       group haproxy
```
I copied the above settings from /etc/haproxy/haproxy.cfg.

Next I had to install and start rsyslog:
```
apt-get install rsyslog -y
/etc/init.d/rsyslog start
```

**Comment by iilyak** [comment]
I think it should use bind instead of COPY. In this case it would always use the most recent version of the source. Also in this case you wouldn't need to rebuild docker image if you update the code. Basically whenever you update the code you would restart docker container instead for rebuilding the image.

**Comment by klaemo** [comment]
@iilyak I see what you're getting at, but you would need to recompile couchdb anyways when you change the source code. That's what the image does. 
There are a few different use cases for the docker image.

**Comment by iilyak** [comment]
I see following use cases for docker images:
1. Use docker to build couchdb on systems without building tools.
2. Use docker to run single node couchdb for testing.
3. Use docker to run cluster of couchdb nodes for testing.
4. Use docker to run couchdb in production.

I think `#4` is out of the scope.
For `#1` you would do
1.1. Write a build script (`dev/docker/build`) which does the build and COPY it into the image. Update PATH so build command could be found.
1.2. bind current directory to the docker container under any known location (`/usr/src/couchdb`) 
1.3. pass name of the command to docker (`build` in this case).
I.e. `docker run -it -name 'db1' -d `pwd`:/usr/src/couchdb docker-image build`
 

For `#2` you would do
```
docker run -it -name 'db1' -e "name=db1" -d `pwd`/dev/data:/data -d `pwd`:/usr/src/couchdb  -p 15984:5984 15986:5986 docker-image run
OR
docker run -it -name 'db1' -d `pwd`/dev/data:/data -d `pwd`:/usr/src/couchdb -p 15984:5984 15986:5986 docker-image run db1
```

For `#3` you could use `https://docs.docker.com/compose/`
OR you could use links manually 

Since the commands are long we could have a helper CLI like:

```
dev/docker-cli run db1
dev/docker-cli run db2
dev/docker-cli run db3
dev/docker-cli run lb
dev/docker-cli run cluster # --- starts 3 nodes cluster
dev/docker-cli run cluster 10 # --- starts 10 nodes cluster
dev/docker-cli build
dev/docker-cli test
```

**Comment by iilyak** [comment]
It is just an idea.

**Comment by klaemo** [comment]
You make some good points. :) I'll give it some more thought. I'm currently on a short trip and don't have my computer with me, so I'll get back to it next week.

**Comment by klaemo** [comment]
I simplified the Dockerfile. It is for local development and testing only. Please give it a spin.

**Comment by iilyak** [comment]
Please consider adding `.dockerignore` to ignore `dev/data` and other bulky things we don't need. Otherwise copying context on `docker build` would take longer.

**Comment by klaemo** [comment]
What would be such things? I'm not familiar with the couchdb code base.

**Comment by iilyak** [comment]
You can use .gitignore as a starting point for .dockerignore.

File: Dockerfile [hunk 1]
```diff
@@ -1,52 +1,52 @@
...
-# Expose nodes on external network interface
-RUN sed -i'' 's/bind_address = 127.0.0.1/bind_address = 0.0.0.0/' rel/overlay/etc/default.ini
+# permissions
+RUN chmod +x /usr/src/couchdb/dev/run && chown -R couchdb:couchdb /usr/src/couchdb
 
-# Build
-RUN ./configure
-RUN make couch
+USER couchdb
+EXPOSE 5984 15984 25984 35984 15986 25986 35986
```
Comment (iilyak) [inline Dockerfile]: There is no way to `remsh` with this list of ports. Unless you add extra ports in command line or use `nsenter`.

Comment (klaemo) [inline Dockerfile]: You can `docker exec` or `docker attach` into the container.

**Comment by klaemo** [comment]
As suggested by @iilyak I added a docker helper script (`dev/docker-cli`), making it easier to orchestrate the docker build. It does the following at the moment:

```
# builds the docker image and then CouchDB 
# inside of the corresponding container with "make couch"
dev/docker-cli build

# ...it also takes ./configure flags as arguments, for example:
dev/docker-cli build --skip-deps

# to build everything run:
dev/docker-cli build all

# to start a cluster with dev/run inside of a container:
dev/docker-cli run

# it passes its arguments to dev/run, for example:
dev/docker-cli run --with-haproxy
```

Let me know, if this makes sense!

**Comment by kxepal** [comment]
@klaemo Nice! 

Btw, have you thought about to update/change make commands? We have there some [commands](https://github.com/apache/couchdb/blob/master/Makefile#L171-L186) to operate with docker. 

I think that we need a) remove these or b) upgrade for docker-cli in order to have one make command to rule everything around.

What do you think?

**Comment by klaemo** [comment]
@kxepal Oh, I wasn't aware of those. I think I'll go with option b) and upgrade them to use docker-cli.

**Comment by klaemo** [comment]
Updated the Makefile to make use of `dev/docker-cli` for the following commands:
```
make docker-image
make docker-start
make docker-clean
```

@kxepal what do you think? :)

**Comment by kxepal** [comment]
@klaemo Looks good, but why rename docker-stop to docker-clean? That assumes that I can start container, but cannot stop it (: Or docker-cli cannot do this?

**Comment by klaemo** [comment]
@kxepal docker-cli starts the container in the foreground (not daemonized). So, once you exit out, it is stopped. It is a change from the previous behavior, but I think it makes sense.

**Comment by kxepal** [comment]
Ah, then it's fine. +1 for me, however, I cannot check this ): @iilyak @janl ?

File: Dockerfile (hunk 1 above)
Comment (iilyak) [inline Dockerfile]: True but how you would do `observer:start()`?

File: dev/docker-cli [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+fi
+
+if [ "$1" = 'run' ]; then
+  printf "Starting CouchDB inside of docker container. Please be patient...\n\n"
+
+  # create container with appropriate port mappings
+  COUCH_CONTAINER_ID="$(docker create -it \
+                  -p 5984:5984 -p 15984:15984 -p 25984:25984 -p 35984:35984 \
+                  -p 15986:15986 -p 25986:25986 -p 35986:35986 \
+                  couchdb-master dev/run ${@:2})"
```
Comment (iilyak) [inline dev/docker-cli]: Would you mind adding `-p 4369:4369 -p 9100-9110:9100-9110"` for remsh to work?

File: dev/docker-cli [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+
+set -e
+
+if [ "$1" = 'build' ]; then
```
Comment (iilyak) [inline dev/docker-cli]: could you add `docker-cli help`?

File: dev/docker-cli (hunk 2 above)
Comment (iilyak) [inline dev/docker-cli]: If you would support remsh it would be nice if you add `docker-cli remsh` command.

File: dev/docker-cli (hunk 3 above)
Comment (klaemo) [inline dev/docker-cli]: good idea. I'll to do that

File: dev/docker-cli (hunk 2 above)
Comment (klaemo) [inline dev/docker-cli]: I added the ports. I'm not familiar with `remsh`. :-/ I'm glad to add such a command, if you tell me what it needs to do :)

**Comment by wohali** [comment]
Hi @klaemo is this still a relevant PR?

**Comment by klaemo** [comment]
I don't think so. There doesn't seem that much interest in a dockerized CouchDB development environment in the community. Also I don't have the time to maintain it anymore.

**Comment by wohali** [comment]
@klaemo Thanks. FYI we _are_ using Docker to build CouchDB in our Apache Jenkins CI environment. Details are over at the https://github.com/apache/couchdb-ci repository. Any comments you have are most welcome :)

Closing this one out.

**Comment by gabrielmancini** [comment]
i use in my local development, but i dont find de logs

**Comment by nono** [comment]
We are using the docker images for couchdb 2.0 for developing https://github.com/cozy/cozy-stack. Thanks you @klaemo!

**Comment by wohali** [comment]
@nono the docker images are unrelated to these changes. Those are over here: https://github.com/klaemo/docker-couchdb

**Pull request closed without merging**